09_Hyper-parameter_tuning_using_Ray/reference_solution/GPT-neo-ray-tune.py

```diff
@@ -0,0 +1,184 @@
+        return tokenizer(
+            texts,
+            max_length=max_length,
+            truncation=True,
+            add_special_tokens=True,
+            return_overflowing_tokens=True,
+            return_length=False,
+        )
+
+    train_dataset_tok = train_dataset.map(
+        tokenize,
+        remove_columns=["text", "label"],
+        batched=True,
+        batch_size=training_args.train_batch_size,
+        num_proc=training_args.dataloader_num_workers,
+    )
+
+    eval_dataset_tok = eval_dataset.map(
+        tokenize,
+        remove_columns=["text", "label"],
+        batched=True,
+        num_proc=training_args.dataloader_num_workers,
+    )
+
+    train_validate_splits = train_dataset_tok.train_test_split(
+        test_size=1000, seed=42, keep_in_memory=True
+    )
+    train_dataset_tok = train_validate_splits["train"]
+    validate_dataset_tok = train_validate_splits["test"]
+
+    collator = DataCollatorForLanguageModeling(
+        tokenizer, mlm=False, return_tensors="pt"
+    )
+
+    trainer = Trainer(
+        model=model,
+        args=training_args,
+        tokenizer=tokenizer,
+        data_collator=collator,
+        train_dataset=train_dataset_tok,
+        eval_dataset=validate_dataset_tok,
+    )
+
+    trainer.train(resume_from_checkpoint=args.resume)
+
+    # report results back to ray
+    eval_results = trainer.evaluate()
+    train.report(
+        dict(
+            loss=eval_results["eval_loss"],
+            perplexity=math.exp(eval_results["eval_loss"]),
+        )
+    )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--model-name",
+        type=str,
+        default="gpt-imdb-model",
+        help="A name for the trained model under. A subdirectory with the given name will be created under the `output-path`.",
+    )
+    parser.add_argument(
+        "--output-path",
+        type=str,
+        help="The root directory under which model checkpoints are stored.",
+    )
+    parser.add_argument(
+        "--logging-path",
+        type=str,
+        help="The root directory under which logging data (for tensorboard) are stored.",
+    )
+    parser.add_argument(
+        "--num-workers",
+        type=int,
+        default=1,
+        help="The number of CPU worker processes to use.",
+    )
+    parser.add_argument(
+        "--resume",
+        default=False,
+        action="store_true",
+        help="If set, continue from a previously interrupted run. Otherwise, overwrite existing checkpoints.",
+    )
+    args, _ = parser.parse_known_args()
+
+    # We need to manually set the number of CPUs and GPUs. Othewise, ray tries to use the whole node and crashes.
+    ray.init(num_cpus=56, num_gpus=8, log_to_driver=False)
+
+    config = {
+        "learning_rate": tune.uniform(
+            1e-6, 1e-3
+        ),  # define search space for learning_rate
+        "args": args,
+    }
+
+    analysis = tune.run(
```

Comment:
Having run the example and watched the output of `rocm-smi`, I feel fairly confident that ray manages to use all the GCDs correctly. The bindings are probably wrong and the multiple GCDs per trial is still an open question. However, I think this can be left as "future work".